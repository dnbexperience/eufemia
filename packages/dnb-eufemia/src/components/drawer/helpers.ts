import { processChildren } from '../../shared/component-helper'

export const checkMinMaxWidth = (
  min: number | string,
  max: number | string,
) => {
  // ensure the min/max don't conflict
  let maxWidth = max
  let minWidth = min
  if (min && !max && parseFloat(String(min)) > 0) {
    maxWidth = 0
  } else if (max && !min && parseFloat(String(max)) > 0) {
    minWidth = 0
  }
  return { minWidth, maxWidth }
}

export const getContent = (props) => {
  if (typeof props.modalContent === 'string') {
    return props.modalContent
  } else if (typeof props.modalContent === 'function') {
    return props.modalContent(props)
  }
  return processChildren(props)
}
