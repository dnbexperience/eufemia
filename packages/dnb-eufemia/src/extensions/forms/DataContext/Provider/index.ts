export { default } from './Provider'
export * from './Provider'
