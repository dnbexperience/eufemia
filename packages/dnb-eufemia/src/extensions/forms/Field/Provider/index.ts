export { default } from './FieldProvider'
export * from './FieldProvider'
