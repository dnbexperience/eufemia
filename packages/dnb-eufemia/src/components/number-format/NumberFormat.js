/**
 * Web NumberFormat Component
 *
 * This is a legacy component.
 * For refferencing while developing new features, please use a Functional component.
 */

import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import Context from '../../shared/Context'
import {
  warn,
  isTrue,
  makeUniqueId,
  validateDOMAttributes,
  convertJsxToString,
  extendPropsWithContextInClassComponent,
  extendDeep,
  detectOutsideClick,
} from '../../shared/component-helper'
import { hasSelectedText, IS_IOS } from '../../shared/helpers'
import {
  spacingPropTypes,
  createSpacingClasses,
} from '../space/SpacingHelper'
import {
  skeletonDOMAttributes,
  createSkeletonClass,
} from '../skeleton/SkeletonHelper'
import Tooltip, { injectTooltipSemantic } from '../tooltip/Tooltip'
import {
  format,
  showSelectionNotice,
  runIOSSelectionFix,
} from './NumberUtils'

export default class NumberFormat extends React.PureComponent {
  static contextType = Context

  static propTypes = {
    id: PropTypes.string,
    value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    locale: PropTypes.string,
    prefix: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
    suffix: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),

    // currency
    currency: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    currency_display: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.oneOf(['code', 'name', 'symbol', 'narrowSymbol', '']),
    ]),
    currency_position: PropTypes.oneOf(['auto', 'before', 'after']),

    // shortens any number or currency including an abbreviation
    compact: PropTypes.oneOfType([
      PropTypes.oneOf(['short', 'long']),
      PropTypes.bool,
    ]),

    // bank account number
    ban: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),

    // national identification number
    nin: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),

    // phone number
    phone: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),

    // organization number
    org: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),

    // percentage
    percent: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),

    // can be tel or sms
    link: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),

    options: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),
    decimals: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    selectall: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    always_selectall: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.bool,
    ]),
    copy_selection: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.bool,
    ]),
    clean_copy_value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.bool,
    ]),
    omit_rounding: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    clean: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    srLabel: PropTypes.node,
    element: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    tooltip: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.func,
      PropTypes.node,
    ]),
    skeleton: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),

    ...spacingPropTypes,

    class: PropTypes.string,
    className: PropTypes.string,
    children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  }
  static defaultProps = {
    id: null,
    value: null,
    locale: null,
    prefix: null,
    suffix: null,
    currency: null,
    currency_display: null, // code, name, symbol
    currency_position: null, // null, before, after
    compact: null,
    ban: null,
    nin: null,
    phone: null,
    org: null,
    percent: null,
    link: null,
    options: null,
    decimals: null,
    selectall: true,
    always_selectall: false,
    copy_selection: true,
    clean_copy_value: false,
    omit_rounding: null,
    clean: null,
    srLabel: null,
    element: 'span', // span or abbr
    tooltip: null,
    skeleton: null,
    class: null,

    className: null,
    children: null,
  }

  constructor(props) {
    super(props)
    this._ref = React.createRef()
    this._selectionRef = React.createRef()

    this._id = props.tooltip ? props.id || makeUniqueId() : undefined
    this.state = { selected: false, omitCurrencySign: false, hover: false }
  }

  componentDidMount() {
    clearTimeout(this._selectAllTimeout)

    // NB: This hack may be removed in future iOS versions
    // in order that iOS v13 can select something on the first try, we run this add range trick
    if (IS_IOS && !hasiOSFix) {
      hasiOSFix = true
      runIOSSelectionFix()
    }
  }

  shortcutHandler = () => {
    showSelectionNotice({
      value: this.cleanedValue,
      label: this.context.getTranslation(this.props)?.NumberFormat
        ?.clipboard_copy,
    }).run(this._ref.current)
  }

  onBlurHandler = () => {
    this.setState({ selected: false })
  }

  onContextMenuHandler = () => {
    if (!hasSelectedText()) {
      clearTimeout(this._selectAllTimeout)
      this._selectAllTimeout = setTimeout(() => {
        this.setFocus()
      }, 1)
    }
  }

  onClickHandler = () => {
    if (
      (isTrue(this.props.selectall) ||
        isTrue(this.props.always_selectall)) &&
      !hasSelectedText()
    ) {
      this.setFocus()
    }
  }

  componentWillUnmount() {
    this.outsideClick?.remove()
  }

  setFocus() {
    this.setState({ selected: true }, () => {
      this._selectionRef.current?.focus()
      this.selectAll()

      if (!isTrue(this.props.copy_selection)) {
        this.outsideClick = detectOutsideClick(
          this._ref.current,
          this.onBlurHandler
        )
      }
    })
  }

  selectAll() {
    try {
      const elem = this._selectionRef.current || this._ref.current
      if (elem) {
        const selection = window.getSelection()
        const range = document.createRange()
        range.selectNodeContents(elem)
        selection.removeAllRanges()
        selection.addRange(range)
      }
    } catch (e) {
      warn(e)
    }
  }

  runFix(comp, className) {
    if (typeof comp === 'function') {
      comp = comp()
    }
    if (React.isValidElement(comp)) {
      return React.cloneElement(comp, {
        className: classnames(comp.props.className, className),
      })
    }
    return <span className={className}>{comp}</span>
  }

  onMouseEnter = () => {
    this.setState({ hover: true })
  }

  onMouseLeave = () => {
    this.setState({ hover: false })
  }

  render() {
    // consume the global context
    const props = extendPropsWithContextInClassComponent(
      this.props,
      NumberFormat.defaultProps,
      this.context.getTranslation(this.props).NumberFormat,
      this.context.NumberFormat
    )

    const {
      id, // eslint-disable-line
      value: _value,
      prefix,
      suffix,
      children,
      currency,
      currency_display,
      currency_position,
      compact,
      ban,
      nin,
      phone,
      org,
      percent,
      link: _link,
      tooltip,
      skeleton,
      options,
      locale,
      decimals,
      omit_rounding,
      clean,
      selectall,
      copy_selection,
      clean_copy_value,
      srLabel,
      element,
      className,
      ..._rest
    } = props
    let rest = _rest

    let link = _link
    let value = _value

    if (value === null && children !== null) {
      value = children
    }

    const formatOptions = {
      locale,
      currency,
      currency_display,
      currency_position,
      omit_currency_sign: this.state.omitCurrencySign,
      compact,
      ban,
      nin,
      phone,
      org,
      percent,
      decimals,
      omit_rounding: isTrue(omit_rounding),
      options,
      clean: isTrue(clean),
      clean_copy_value: isTrue(clean_copy_value),
      returnAria: true,
    }

    // use only the props from context, who are available here anyway
    const useContext = extendDeep(
      { locale: null, currency: null },
      this.context
    )

    if (useContext) {
      if (useContext.locale && !locale) {
        formatOptions.locale = useContext.locale
      }

      // only replace if the prop is "true" and not actually a currency
      if (useContext.currency && isTrue(currency)) {
        formatOptions.options = formatOptions.options
          ? { ...formatOptions.options }
          : {}
        formatOptions.options.currency = useContext.currency
      }
    }

    let {
      cleanedValue,
      number: display,
      aria,
      locale: lang,
    } = format(value, formatOptions)
    this.cleanedValue = cleanedValue

    if (prefix) {
      display = (
        <>
          {this.runFix(prefix, 'dnb-number-format__prefix')} {display}
        </>
      )
      aria = String(
        `${convertJsxToString(
          this.runFix(prefix, 'dnb-number-format__prefix')
        )} ${aria}`
      )
    }
    if (suffix) {
      display = (
        <>
          {display} {this.runFix(suffix, 'dnb-number-format__suffix')}
        </>
      )
      aria = `${aria} ${convertJsxToString(
        this.runFix(suffix, 'dnb-number-format__suffix')
      )}`
    }

    if (tooltip) {
      rest = injectTooltipSemantic(rest)
    }

    const attributes = {
      lang,
      ref: this._ref,
      className: classnames(
        'dnb-number-format',
        className,
        (isTrue(currency) || typeof currency === 'string') &&
          'dnb-number-format--currency',
        isTrue(selectall) && 'dnb-number-format--selectall',
        this.state.selected && 'dnb-number-format--selected',
        link && 'dnb-anchor',
        createSpacingClasses(this.props)
      ),

      // Makes it possible for NVDA to read on mouse over
      onMouseEnter: this.onMouseEnter,
      onMouseLeave: this.onMouseLeave,

      ...rest,
    }

    /**
     * Works in VoiceOver and NVDA
     * Makes the span with it's roles etc. appear as text.
     * Special useful if a number is in side e.g. a paragraph alongside with numbers
     */
    attributes['role'] = 'text' // role="text"

    const displayParams = {}
    if (isTrue(selectall) || isTrue(copy_selection)) {
      displayParams.onClick = this.onClickHandler
      displayParams.onContextMenu = this.onContextMenuHandler
    }

    validateDOMAttributes(this.props, attributes)
    skeletonDOMAttributes(attributes, skeleton, this.context)

    if (link) {
      if (isTrue(link)) {
        link = 'tel'
      }
      return (
        <a href={`${link}:${display}`} {...attributes}>
          {display}
        </a>
      )
    }

    const Element = element

    return (
      <Element {...attributes}>
        <span
          className={classnames(
            'dnb-number-format__visible',
            createSkeletonClass('font', skeleton, this.context)
          )}
          // Makes it possible for NVDA to read on mouse over
          aria-hidden={!this.state.hover}
          {...displayParams}
        >
          {display}
        </span>

        {/* Used for VoiceOver and NVDA when navigating with arrow keys */}
        <span
          className="dnb-sr-only"
          // Use "data-text" so Chrome does not copy the HTML as content, when pasting it in Outlook etc.
          data-text={
            srLabel ? `${convertJsxToString(srLabel)}${' '}${aria}` : aria
          }
        />

        {isTrue(copy_selection) && (
          <span
            className="dnb-number-format__selection dnb-no-focus"
            ref={this._selectionRef}
            tabIndex={-1}
            onBlur={this.onBlurHandler}
            onCopy={this.shortcutHandler}
            aria-hidden
          >
            {this.state.selected && cleanedValue}
          </span>
        )}

        {tooltip && (
          <Tooltip
            id={this._id + '-tooltip'}
            targetElement={this._ref}
            tooltip={tooltip}
          />
        )}
      </Element>
    )
  }
}

let hasiOSFix = false

NumberFormat._supportsSpacingProps = true
