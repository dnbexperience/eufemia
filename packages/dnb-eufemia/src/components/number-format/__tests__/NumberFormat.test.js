/**
 * Component Test
 *
 */

import React from 'react'
import {
  mount,
  axeComponent,
  toJson,
  loadScss,
  mockGetSelection,
} from '../../../core/jest/jestSetup'
import { LOCALE } from '../../../shared/defaults'
import { isMac } from '../../../shared/helpers'
import Provider from '../../../shared/Provider'
import NumberFormat from '../NumberFormat'
import { format } from '../NumberUtils'

const Component = (props) => {
  return <NumberFormat id="unique" {...props} />
}

const element = NumberFormat.defaultProps.element
const locale = LOCALE
const value = 12345678.9876
const snapshotProps = {
  value,
  locale,
  element,
}

// make it possible to change the navigator lang
// because "navigator.language" defaults to en-GB
let languageGetter, platformGetter

beforeAll(() => {
  languageGetter = jest.spyOn(window.navigator, 'language', 'get')
  platformGetter = jest.spyOn(window.navigator, 'platform', 'get')

  // simulate mac, has to run on the first render
  platformGetter.mockReturnValue('Mac')
  languageGetter.mockReturnValue(locale)

  isMac() // just to update the exported const: IS_MAC

  mockGetSelection()
})

describe('NumberFormat component', () => {
  const displaySelector = element + '.dnb-number-format span'
  const ariaSelector = element + '.dnb-number-format span[id]'

  it('have to match default number-format snapshot', () => {
    const Comp = mount(<Component {...snapshotProps} />)
    expect(toJson(Comp)).toMatchSnapshot()
  })

  it('have to match default number', () => {
    const Comp = mount(<Component value={value} />)
    expect(Comp.find(displaySelector).first().text()).toBe(
      '12 345 678,9876'
    )
  })

  it('have to match currency for default locale', () => {
    const Comp = mount(<Component value={-value} currency />)

    expect(Comp.find(displaySelector).first().text()).toBe(
      '-12 345 678,99 kr'
    )

    expect(Comp.find(ariaSelector).first().text()).toBe(
      '-12 345 678,99 norske kroner'
    )

    // also check the formatting with one digit less
    Comp.setProps({
      children: null,
      decimals: 0,
      value: 12345,
    })

    expect(Comp.find(displaySelector).first().text()).toBe('12 345 kr')
  })

  it('have to match currency in en locale', () => {
    const Comp = mount(<Component value={-value} currency locale="en" />)

    expect(Comp.find(displaySelector).first().text()).toBe(
      '-NOK 12 345 678.99'
    )

    expect(Comp.find(ariaSelector).first().text()).toBe(
      '-12 345 678.99 Norwegian kroner'
    )

    // also check the formatting with one digit less
    Comp.setProps({
      children: null,
      decimals: 0,
      value: 12345,
    })

    expect(Comp.find(displaySelector).first().text()).toBe('NOK 12 345')
  })

  it('have to match currency with large decimals', () => {
    const Comp = mount(<Component value="5000.0099" currency />)
    expect(Comp.find(displaySelector).first().text()).toBe('5 000,01 kr')
  })

  it('has valid selected number', () => {
    const Comp = mount(<Component value={-value} currency />)

    const selection = window.getSelection()
    selection.removeAllRanges()

    expect(
      Comp.find('span').first().hasClass('dnb-number-format--selected')
    ).toBe(false)

    Comp.find('.dnb-number-format__visible').simulate('click')

    expect(
      Comp.find('span').first().hasClass('dnb-number-format--selected')
    ).toBe(true)

    const { cleanedValue: noVal } = format(-value, {
      currency: true,
      returnAria: true,
    })
    expect(Comp.find('.dnb-number-format__selection').text()).toBe(noVal)
    expect(window.getSelection().toString()).toBe('1234.56') // Hack! Having there the "cleanedNumber" would be optimal.
    expect(window.getSelection().rangeCount).toBe(1)

    Comp.setProps({ locale: 'en-GB' })
    const { cleanedValue: enVal } = format(-value, {
      locale: 'en-GB',
      currency: true,
      returnAria: true,
    })
    expect(Comp.find('.dnb-number-format__selection').text()).toBe(enVal)
  })

  it('have to match currency with currency_position="after"', () => {
    const Comp = mount(
      <Component value={-value} currency currency_position="after" />
    )

    expect(Comp.find(displaySelector).first().text()).toBe(
      '-12 345 678,99 kr'
    )

    expect(Comp.find(ariaSelector).first().text()).toBe(
      '-12 345 678,99 norske kroner'
    )

    Comp.setProps({
      locale: 'en-GB',
    })

    expect(Comp.find(displaySelector).first().text()).toBe(
      '-12 345 678.99 NOK'
    )

    expect(Comp.find(ariaSelector).first().text()).toBe(
      '-12 345 678.99 Norwegian kroner'
    )

    Comp.setProps({
      currency_display: 'code',
    })
    expect(Comp.find(displaySelector).first().text()).toBe(
      '-12 345 678.99 NOK'
    )

    Comp.setProps({
      locale: 'no',
    })

    Comp.setProps({
      currency_position: 'before',
    })
    expect(Comp.find(displaySelector).first().text()).toBe(
      'NOK -12 345 678,99'
    )
  })

  it('have to match currency with currency_position="before"', () => {
    const Comp = mount(
      <Component value={-value} currency currency_position="before" />
    )

    expect(Comp.find(displaySelector).first().text()).toBe(
      'kr -12 345 678,99'
    )

    expect(Comp.find(ariaSelector).first().text()).toBe(
      '-12 345 678,99 norske kroner'
    )

    Comp.setProps({
      locale: 'en-GB',
    })

    expect(Comp.find(displaySelector).first().text()).toBe(
      '-NOK 12 345 678.99'
    )

    expect(Comp.find(ariaSelector).first().text()).toBe(
      '-12 345 678.99 Norwegian kroner'
    )

    Comp.setProps({
      currency_display: 'code',
    })
    expect(Comp.find(displaySelector).first().text()).toBe(
      '-NOK 12 345 678.99'
    )

    Comp.setProps({
      locale: 'no',
    })

    Comp.setProps({
      currency_position: 'after',
    })
    expect(Comp.find(displaySelector).first().text()).toBe(
      '-12 345 678,99 NOK'
    )
  })

  it('have to match currency under 100.000', () => {
    const Comp = mount(<Component value={-12345.95} currency />)

    expect(Comp.find(displaySelector).first().text()).toBe('-12 345,95 kr')

    expect(Comp.find(ariaSelector).first().text()).toBe(
      '-12345,95 norske kroner'
    )
  })

  it('have to match currency with no decimals', () => {
    const Comp = mount(
      <Component value={-12345.99} currency decimals={0} />
    )

    expect(Comp.find(displaySelector).first().text()).toBe('-12 346 kr')

    expect(Comp.find(ariaSelector).first().text()).toBe(
      '-12346 norske kroner'
    )
  })

  it('have to match phone number', () => {
    const Comp = mount(<Component phone>+47 99999999</Component>)
    expect(Comp.find(displaySelector).first().text()).toBe(
      '0047 99 99 99 99'
    )
  })

  it('have to match bank account number', () => {
    const Comp = mount(<Component ban>20001234567</Component>)
    expect(Comp.find(displaySelector).first().text()).toBe('2000 12 34567')

    // also check the formatting with one digit less
    Comp.setProps({
      children: null,
      value: 2000123456,
    })
    expect(Comp.find(displaySelector).first().text()).toBe('2000 12 3456')
  })

  it('have to match national identification number', () => {
    const Comp = mount(<Component nin>18089212345</Component>)
    expect(Comp.find(displaySelector).first().text()).toBe('180892 12345')
    expect(Comp.find(ariaSelector).first().text()).toBe(
      '18 08 92 1 2 3 4 5'
    )
  })

  it('have to match organization number', () => {
    const Comp = mount(
      <Component org suffix="MVA">
        123456789
      </Component>
    )
    expect(Comp.find(displaySelector).first().text()).toBe(
      '123 456 789 MVA'
    )
    expect(Comp.find(ariaSelector).first().text()).toBe(
      '1 2 3 4 5 6 7 8 9 MVA'
    )
  })

  it('have to handle prefix and suffix', () => {
    const Comp = mount(
      <Component prefix={<span>prefix</span>} suffix={<span>suffix</span>}>
        123456789.5
      </Component>
    )
    expect(Comp.find(displaySelector).first().text()).toBe(
      'prefix 123 456 789,5 suffix'
    )
    expect(Comp.find(ariaSelector).first().text()).toBe(
      'prefix 123 456 789,5 suffix'
    )
  })

  it('will add visually hidden label when srLabel is given', () => {
    const Comp = mount(
      <Component value={-value} currency srLabel="Total:" />
    )
    expect(Comp.find('.dnb-sr-only').first().text()).toBe('Total: ')
    expect(Comp.text()).toContain('Total: -12 345 678,99 kr')
  })

  it('should validate with ARIA rules', async () => {
    const Comp = mount(
      <Component value={-value} currency srLabel="Total:" />
    )
    expect(await axeComponent(Comp)).toHaveNoViolations()
  })
})

describe('NumberFormat compact', () => {
  const displaySelector = element + '.dnb-number-format span'
  const ariaSelector = element + '.dnb-number-format span[id]'

  it('have to match default compact number', () => {
    const Comp = mount(<Component value={-value} compact decimals={1} />)
    expect(Comp.find(displaySelector).first().text()).toBe('-12,3 mill.')
    expect(Comp.find(ariaSelector).first().text()).toBe('-12,3 millioner')
  })

  it('have to match short compact number', () => {
    const Comp = mount(
      <Component value={-12345} compact="short" decimals={3} />
    )
    expect(Comp.find(displaySelector).first().text()).toBe('-12,345k')
    expect(Comp.find(ariaSelector).first().text()).toBe('-12,345 tusen')
  })

  it('have to match long compact number', () => {
    const Comp = mount(
      <Component value={-value} compact="long" decimals={3} />
    )
    expect(Comp.find(displaySelector).first().text()).toBe(
      '-12,346 millioner'
    )
    expect(Comp.find(ariaSelector).first().text()).toBe(
      '-12,346 millioner'
    )
  })

  it('have to match currency based compact number', () => {
    const Comp = mount(
      <Component value={-value} compact currency decimals={2} />
    )
    expect(Comp.find(displaySelector).first().text()).toBe(
      '-12,35 mill. kr'
    )
    expect(Comp.find(ariaSelector).first().text()).toBe(
      '-12,35 millioner norske kroner'
    )
  })

  it('have to match currency based compact number with custom currency_display', () => {
    const Comp = mount(
      <Component
        compact="long"
        currency
        value={-value}
        decimals={3}
        currency_display="name"
      />
    )
    expect(Comp.find(displaySelector).first().text()).toBe(
      '-12,346 millioner norske kroner'
    )
    expect(Comp.find(ariaSelector).first().text()).toBe(
      '-12,346 millioner norske kroner'
    )
  })

  it('have to match compact number with custom decimals', () => {
    const Comp = mount(
      <Component value={-value} compact currency decimals={4} />
    )
    expect(Comp.find(displaySelector).first().text()).toBe(
      '-12,3457 mill. kr'
    )
    expect(Comp.find(ariaSelector).first().text()).toBe(
      '-12,3457 millioner norske kroner'
    )
  })

  describe('en-GB', () => {
    it('have to match default compact number', () => {
      const Comp = mount(
        <Component value={-value} compact locale="en-GB" decimals="2" />
      )
      expect(Comp.find(displaySelector).first().text()).toBe('-12.35M')
      expect(Comp.find(ariaSelector).first().text()).toBe('-12.35 million')
    })

    it('have to match long compact number', () => {
      const Comp = mount(
        <Component
          value={-value}
          compact="long"
          locale="en-GB"
          decimals="2"
        />
      )
      expect(Comp.find(displaySelector).first().text()).toBe(
        '-12.35 million'
      )
      expect(Comp.find(ariaSelector).first().text()).toBe('-12.35 million')
    })

    it('have to match currency based compact number', () => {
      const Comp = mount(
        <Component
          value={-value}
          compact
          currency
          locale="en-GB"
          decimals="2"
        />
      )
      expect(Comp.find(displaySelector).first().text()).toBe('-NOK 12.35M')
      expect(Comp.find(ariaSelector).first().text()).toBe(
        '-12.35 million Norwegian kroner'
      )
    })
  })

  const numbersDecimals = [
    {
      value: -123,
      display: '-123,0',
      aria: '-123,0',
    },
    {
      value: -1234,
      display: '-1,2 tusen',
      aria: '-1,2 tusen',
    },
    {
      value: -12345,
      display: '-12,3 tusen',
      aria: '-12,3 tusen',
    },
    {
      value: -123456,
      display: '-123,5 tusen',
      aria: '-123,5 tusen',
    },
    {
      value: -1234567,
      display: '-1,2 mill.',
      aria: '-1,2 millioner',
    },
    {
      value: -12345678,
      display: '-12,3 mill.',
      aria: '-12,3 millioner',
    },
    {
      value: -123456789,
      display: '-123,5 mill.',
      aria: '-123,5 millioner',
    },
    {
      value: -1234567891,
      display: '-1,2 mrd.',
      aria: '-1,2 milliarder',
    },
    {
      value: -12345678912,
      display: '-12,3 mrd.',
      aria: '-12,3 milliarder',
    },
    {
      value: -123456789123,
      display: '-123,5 mrd.',
      aria: '-123,5 milliarder',
    },
    {
      value: -1234567891234,
      display: '-1,2 bill.',
      aria: '-1,2 billioner',
    },
    {
      value: -12345678912345,
      display: '-12,3 bill.',
      aria: '-12,3 billioner',
    },
    {
      value: -123456789123456,
      display: '-123,5 bill.',
      aria: '-123,5 billioner',
    },
  ]

  it.each(numbersDecimals)(
    'have to match compact number %s',
    ({ value, display, aria }) => {
      const Comp = mount(<Component value={value} compact decimals={1} />)

      expect(Comp.find(displaySelector).first().text()).toBe(display)
      expect(Comp.find(ariaSelector).first().text()).toBe(aria)
    }
  )
})

describe('NumberFormat component with provider', () => {
  const displaySelector = element + '.dnb-number-format span'

  it('have to match inherit properties', () => {
    const Comp = mount(
      <Provider
        locale="en-GB"
        NumberFormat={{ currency: true, currency_display: 'name' }}
      >
        <Component value={value} />
      </Provider>
    )
    expect(Comp.find(displaySelector).first().text()).toBe(
      '12 345 678.99 Norwegian kroner'
    )
  })
})

describe('NumberFormat scss', () => {
  it('have to match snapshot', () => {
    const scss = loadScss(require.resolve('../style/deps.scss'))
    expect(scss).toMatchSnapshot()
  })
})
