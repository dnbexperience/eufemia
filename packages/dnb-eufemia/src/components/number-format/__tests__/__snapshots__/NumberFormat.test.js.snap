// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberFormat component have to match default number-format snapshot 1`] = `
<Component
  element="span"
  locale="nb-NO"
  value={12345678.9876}
>
  <NumberFormat
    always_selectall={false}
    ban={null}
    class={null}
    className={null}
    clean={null}
    clean_copy_value={false}
    compact={null}
    copy_selection={true}
    currency={null}
    currency_display={null}
    currency_position={null}
    decimals={null}
    element="span"
    id="unique"
    link={null}
    locale="nb-NO"
    nin={null}
    omit_rounding={null}
    options={null}
    org={null}
    percent={null}
    phone={null}
    prefix={null}
    selectall={true}
    skeleton={null}
    srLabel={null}
    suffix={null}
    tooltip={null}
    value={12345678.9876}
  >
    <span
      className="dnb-number-format dnb-number-format--selectall"
      lang="nb-NO"
      role="text"
    >
      <span
        aria-describedby="unique"
        aria-hidden={true}
        className="dnb-number-format__visible"
        onClick={[Function]}
        onContextMenu={[Function]}
      >
        12 345 678,9876
      </span>
      <span
        className="dnb-number-format__sr-only dnb-sr-only"
        id="unique"
      >
        12 345 678,9876
      </span>
      <span
        aria-hidden={true}
        className="dnb-number-format__selection dnb-no-focus"
        onBlur={[Function]}
        onCopy={[Function]}
        tabIndex={-1}
      >
        12345678,9876
      </span>
    </span>
  </NumberFormat>
</Component>
`;

exports[`NumberFormat scss have to match snapshot 1`] = `
"/*
* DNB NumberFormat
*
*/
/**
 * This file is only used to make components independent
 * so that they can get imported individually, without the core styles
 *
 */
/*
 * Utilities
 */
/*
 * Scopes
 *
 */
/*
 * Document Reset
 *
 */
/*
* DNB Tooltip
*
*/
.dnb-tooltip {
  font-family: var(--font-family-default);
  font-weight: var(--font-weight-basis);
  font-size: var(--font-size-small);
  font-style: normal;
  line-height: var(--line-height-basis);
  color: var(--color-black-80, #333);
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
  /**
  * Ensure consistency and use the same as HTML reset -> html {...}
  * between base and code package
  */
  -moz-tab-size: 4;
  tab-size: 4;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  word-break: break-word;
  /**
  * 1. Remove repeating backgrounds in all browsers (opinionated).
  * 2. Add border box sizing in all browsers (opinionated).
  */
  /**
  * 1. Add text decoration inheritance in all browsers (opinionated).
  * 2. Add vertical alignment inheritance in all browsers (opinionated).
  */
  margin: 0;
  padding: 0; }
  .dnb-tooltip *,
  .dnb-tooltip ::before,
  .dnb-tooltip ::after {
    background-repeat: no-repeat;
    /* 1 */
    box-sizing: border-box;
    /* 2 */ }
  .dnb-tooltip ::before,
  .dnb-tooltip ::after {
    text-decoration: inherit;
    /* 1 */
    vertical-align: inherit;
    /* 2 */ }

/*
* Tooltip component
*
*/
.dnb-tooltip {
  position: absolute;
  z-index: 3100;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 0 1rem;
  opacity: 0;
  visibility: hidden;
  transition: opacity 200ms var(--easing-default); }
  .dnb-tooltip--large {
    padding: 0.25rem 1rem; }
  .dnb-tooltip--animate_position {
    transition: all 200ms var(--easing-default), opacity 200ms var(--easing-default); }
  .dnb-tooltip--active {
    visibility: visible;
    /*
      because of the first "show" we also use animation
      also, use forwards because of the usage of visibility
    */
    animation: show-tooltip 200ms var(--easing-default) forwards; }
  html[data-visual-test] .dnb-tooltip--active, .dnb-tooltip--active.dnb-tooltip--no-animation {
    animation: show-tooltip 1ms var(--easing-default) forwards; }
  .dnb-tooltip--hide {
    visibility: visible;
    animation: hide-tooltip 200ms var(--easing-default) forwards; }
  .dnb-tooltip--hide.dnb-tooltip--no-animation {
    animation: hide-tooltip 1ms linear forwards; }
  .dnb-tooltip--fixed {
    position: fixed; }
  html[data-visual-test] .dnb-tooltip--hide {
    animation: hide-tooltip 1ms linear 1s forwards; }
  .dnb-tooltip__portal {
    position: absolute;
    top: 0;
    left: 0;
    right: 0; }
  .dnb-tooltip__content {
    min-width: 2rem;
    min-height: 1.5rem;
    padding: 0; }
  .dnb-tooltip__arrow {
    position: absolute;
    pointer-events: none;
    margin: 0;
    width: 1rem;
    height: 0.5rem;
    overflow: hidden; }
    .dnb-tooltip__arrow::before {
      content: '';
      position: absolute;
      bottom: 0;
      left: 0;
      width: 1rem;
      height: 1rem;
      transform: translateY(70%) rotate(45deg); }
    .dnb-tooltip__arrow__position--bottom {
      top: -0.5rem; }
    .dnb-tooltip__arrow__position--top {
      bottom: -0.5rem;
      transform: rotate(180deg); }
    .dnb-tooltip__arrow__position--left {
      right: -0.75rem;
      margin-right: 3px;
      transform: rotate(90deg); }
    .dnb-tooltip__arrow__position--right {
      left: -0.75rem;
      margin-left: 3px;
      transform: rotate(270deg); }
    .dnb-tooltip__arrow__arrow--left {
      align-self: flex-start; }
    .dnb-tooltip__arrow__arrow--right {
      align-self: flex-end; }

@keyframes show-tooltip {
  from {
    opacity: 0; }
  to {
    opacity: 1; } }

@keyframes hide-tooltip {
  from {
    opacity: 1; }
  to {
    opacity: 0;
    visibility: hidden; } }

/*
 * NumberFormat Component
 *
 */
/*
 * Anchor mixins
 *
 */
.dnb-number-format {
  display: inline-block;
  position: relative;
  speak-as: numbers;
  color: inherit;
  white-space: nowrap; }
  .dnb-number-format__prefix, .dnb-number-format__suffix {
    white-space: normal; }
  .dnb-number-format.dnb-anchor {
    cursor: pointer; }
  .dnb-number-format--selectall {
    cursor: text; }
  .dnb-number-format__visible {
    position: relative;
    z-index: 2; }
  .dnb-number-format--selected .dnb-number-format__visible {
    color: var(--color-black);
    background-color: var(--color-mint-green); }
  .dnb-number-format__selection {
    position: absolute;
    z-index: 1;
    top: 0;
    left: 0;
    opacity: 0;
    user-select: none; }
  .dnb-number-format--selected .dnb-number-format__selection {
    z-index: 3;
    user-select: all; }
  .dnb-number-format.dnb-skeleton,
  .dnb-skeleton .dnb-number-format {
    -webkit-text-fill-color: var(--skeleton-color); }
"
`;
